### PR TITLE
Testing & fixes for ranking

### LINKED ISSUE #180: Implement test cases for ranking
See #179, #178, #148

### PR DESCRIPTION
**Pull request by lidavidm**
Crashed bots improperly had `last_turn_alive` updated at the end of the game.

Fixes #180.

### CODE REVIEW

**Comment by snaar** [comment]
Unit tests :D

**Comment by lidavidm** [comment]
Oh, would this be considered a unit test?

File: game_engine/core/HaliteImpl.cpp [hunk 1]
```diff
@@ -590,6 +590,7 @@ void HaliteImpl::kill_player(const Player::id_type &player_id) {
     Logging::log("Killing player", Logging::Level::Warning, player_id);
     auto &player = game.store.get_player(player_id);
     player.terminated = true;
+    player.can_play = false;
     game.networking.kill_player(player);
```
Comment (Janzert) [inline game_engine/core/HaliteImpl.cpp]: The second change seems like it makes the first redundant? Not that it needs to go, just making sure I understand correctly. The tests here also still pass just fine without the change to line 474.

These changes also make a small change to the assertion in the comment at line 477-478 `liveness lasts to the end of a turn` now should continue with `in which a bot makes a valid move`, i.e. when a bot is terminated it is dead immediately and the previous move really was the 'last turn alive'.

One other thing I noticed while looking at this, while it isn't changed here it seems like the shortcut return at line 486 is wrong (e.g. if the first two players are still alive, players 3 and 4 will not be checked or have their stats updated).

Comment (lidavidm) [reply game_engine/core/HaliteImpl.cpp]: Yes, the two changes aren't quite both needed, but I'd rather make sure the checks reflect the intent (only update last_turn_alive for non-terminated players who qualify to play). 

I think sending invalid moves should mean you aren't alive on that turn. Your resources/ships are cleared immediately, so your liveness should as well.

I agree the shortcut return is wrong, I'll remove it.

Comment (lidavidm) [reply game_engine/core/HaliteImpl.cpp]: Unfortunately taking out the shortcut breaks the ranking, will have to dig into it...

Comment (Janzert) [reply game_engine/core/HaliteImpl.cpp]: Yes, this definition for the extent of 'liveness' makes sense to me as well. Sorry I wasn't clear that the return just needed to move not be removed.

Everything looks good to me.

**Pull request merged**